### PR TITLE
Migrate from polyrepo to monorepo

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## **Description**

This PR migrates the current MetaMask design system repo which is currently using the metamask-module-template and a polyrepo structure to a monorepo structure. The primary goal is to streamline the module management process by consolidating multiple repositories into a single monorepo. The `core` module serves as a reference for the files and structure required in the monorepo setup. This change will simplify dependency management and improve development efficiency across modules.

Screenshot 2024-10-08 at 2 09 12 PM

### Key changes:
- Updated module configuration to align with the monorepo structure.
- Consolidated files and dependencies into the monorepo format.
- Ensured consistency with the `core` module in terms of file inclusion, structure, and build processes.

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-design-system/issues/5

## **Manual testing steps**

1. Clone the repository and navigate to the monorepo.
2. Run all `yarn` scripts to ensure they work with the exception of `prepare-preview-build` and `publish-previews`
3. Verify that the scriptsfunction as expected within the monorepo.
4. Ensure the build passes without issues in the monorepo format.
5. Ensure no instances of "core" remain unless links to information for migrating packages

## **Screenshots/Recordings**

Video showing all `yarn` scripts work from `package.json` with the exception of `prepare-preview-build` and `publish-previews` issue here: https://github.com/MetaMask/metamask-design-system/issues/26

https://www.loom.com/share/4f40fa33ae45472e807137d3995c212c

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs)
- [x] I’ve completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and/or screenshots.

### CODE REVIEW

**Comment by georgewrmarshall** [comment]
@SocketSecurity ignore all

**Comment by georgewrmarshall** [comment]
@SocketSecurity ignore-all

File: .github/workflows/publish-preview.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
+name: Publish a preview build
```
Comment (georgewrmarshall) [inline .github/workflows/publish-preview.yml:1]: `Publish a preview build` currently doesn't work, but we'd like to utilize the ability to publish package previews. A recent example is the brand evolution work for the design tokens and publishing builds like in the portfolio. I've created an issue here: https://github.com/MetaMask/metamask-design-system/issues/26.

File: .github/workflows/publish-release.yml [hunk 2]
```diff
@@ -69,53 +77,22 @@ jobs:
...
-  publish-release-to-latest-gh-pages:
-    needs: publish-npm
-    name: Publish docs to `latest` directory of `gh-pages` branch
-    permissions:
-      contents: write
-    uses: ./.github/workflows/publish-docs.yml
-    with:
-      destination_dir: latest
-    secrets:
-      PUBLISH_DOCS_TOKEN: ${{ secrets.PUBLISH_DOCS_TOKEN }}
```
Comment (georgewrmarshall) [inline .github/workflows/publish-release.yml:121]: Aligns to core repo

File: .github/workflows/security-code-scanner.yml [hunk 3]
```diff
@@ -19,25 +19,12 @@ jobs:
...
+            '**/*.test.ts'
             node_modules
-            storybook/
-            test*/
-          rules_excluded: example
-          project_metrics_token: ${{ secrets.SECURITY_SCAN_METRICS_TOKEN }}
+            merged-packages/
+            '**/jest.environment.js'
+          project_metrics_token: ${{secrets.SECURITY_SCAN_METRICS_TOKEN}}
           slack_webhook: ${{ secrets.APPSEC_BOT_SLACK_WEBHOOK }}
```
Comment (georgewrmarshall) [inline .github/workflows/security-code-scanner.yml:30]: Aligns with core

File: .gitignore [hunk 4]
```diff
@@ -76,3 +29,6 @@ node_modules/
 !.yarn/releases
 !.yarn/sdks
 !.yarn/versions
+
+# typescript
+packages/*/*.tsbuildinfo
```
Comment (georgewrmarshall) [inline .gitignore:34]: Aligns with core

File: .prettierrc.js [hunk 5]
```diff
@@ -1,6 +1,9 @@
...
-// for explicitness
+/**
+ * @type {import('prettier').Options}
+ */
 module.exports = {
+  // All of these are defaults except singleQuote, but we specify them
+  // for explicitness
   quoteProps: 'as-needed',
   singleQuote: true,
   tabWidth: 2,
```
Comment (georgewrmarshall) [inline .prettierrc.js:9]: Aligns with core

File: .yarnrc.yml [hunk 6]
```diff
@@ -4,7 +4,7 @@ enableGlobalCache: false
 
 enableScripts: false
 
-enableTelemetry: 0
+enableTelemetry: false
```
Comment (georgewrmarshall) [inline .yarnrc.yml:7]: Hmm this aligns with core but it seems the docs say it should be `0` https://yarnpkg.com/advanced/telemetry. Revert this change it's for improving `yarn`. I'm actually not opposed to this being turned on 🤷

File: README.md [hunk 7]
```diff
@@ -2,94 +2,33 @@
...
-6. Squash & Merge the release.
-
-   - This should trigger the [`action-publish-release`](https://github.com/MetaMask/action-publish-release) workflow to tag the final release commit and publish the release on GitHub.
-
-7. Publish the release on npm.
-
-   - Wait for the `publish-release` GitHub Action workflow to finish. This should trigger a second job (`publish-npm`), which will wait for a run approval by the [`npm publishers`](https://github.com/orgs/MetaMask/teams/npm-publishers) team.
-   - Approve the `publish-npm` job (or ask somebody on the npm publishers team to approve it for you).
-   - Once the `publish-npm` job has finished, check npm to verify that it has been published.
+[^fn1]: The package list and dependency graph should be programmatically generated by running `yarn update-readme-content`.
```
Comment (georgewrmarshall) [inline README.md:34]: This aligns with the core monorepo, and most docs have been moved to the contributor docs. We can further refine it to match our usage once we’ve published a package and established best practices for testing, component creation, Storybook usage, etc.

File: babel.config.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+// We use Babel for our tests in scripts/.
+module.exports = {
+  env: {
+    test: {
+      presets: ['@babel/preset-typescript'],
+      plugins: ['@babel/plugin-transform-modules-commonjs'],
+    },
+  },
+};
```
Comment (georgewrmarshall) [inline babel.config.js:9]: babel is used for the create-package tests that I think we want to use to create new packages

File: docs/contributing.md [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+This repository relies on Yarn's [workspaces feature](https://yarnpkg.com/features/workspaces) to provide a way to work with packages individually and collectively. Refer to the documentation for the following Yarn commands for usage instructions:
+
+- [`yarn workspace`](https://yarnpkg.com/cli/workspace)
+- [`yarn workspaces foreach`](https://yarnpkg.com/cli/workspaces/foreach)
+
+> **Note**
+>
+> - `workspaceName` in the Yarn documentation is the `name` field within a package's `package.json`, e.g., `@metamask/metamask-module-template`, not the directory where it is located, e.g., `packages/metamask-module-template`.
+> - `commandName` in the Yarn documentation is any sub-command that the `yarn` executable would usually take. Pay special attention to the difference between `run` vs `exec`. If you want to run a package script, you would use `run`, e.g., `yarn workspace @metamask/metamask-module-template run changelog:validate`; but if you want to run _any_ shell command, you'd use `exec`, e.g. `yarn workspace @metamask/metamask-module-template exec cat package.json | jq '.version'`.
```
Comment (georgewrmarshall) [inline docs/contributing.md:38]: Created an issue to replace this once we merge our first package https://github.com/MetaMask/metamask-design-system/issues/27

File: docs/contributing.md [hunk 10]
```diff
@@ -0,0 +1,267 @@
+# Contributing to the monorepo
```
Comment (georgewrmarshall) [inline docs/contributing.md:1]: The majority of these contributor docs holds true for the design system repo which is why I'm opting to keep it. There are some instnaces of `metamask-module-template` that need to be replaced once we have merged/shipped our first package. Issue here https://github.com/MetaMask/metamask-design-system/issues/27

File: docs/migrate-tags.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
+# `migrate-tags`
```
Comment (georgewrmarshall) [inline docs/migrate-tags.md:1]: These docs will be helpful for migrating the `design-tokens` and `eslint-plugin-design-tokens` packages, and they also reference a tool to assist with the migrations. Once those packages are migrated, we can decide whether to keep this documentation or remove it, as we likely won’t have any further packages to migrate, unlike core.

File: docs/package-migration-process-guide.md [hunk 12]
```diff
@@ -0,0 +1,246 @@
+# Package Migration Process Guide
```
Comment (georgewrmarshall) [inline docs/package-migration-process-guide.md:1]: More docs about migration that we can use for the design-tokens and eslint-plugin-design-tokens pacakge

File: docs/reviewing-release-prs.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
+# Reviewing Release PRs
```
Comment (georgewrmarshall) [inline docs/reviewing-release-prs.md:1]: Although we’ll likely operate slightly differently from core, with fewer teams contributing, I still think this is a great resource for the release process and something we should keep as we improve at handling releases.

File: jest.config.packages.js [hunk 14]
```diff
@@ -28,25 +28,17 @@ module.exports = {
   coverageDirectory: 'coverage',
```
Comment (georgewrmarshall) [inline jest.config.packages.js:28]: Aligns with core

File: jest.config.scripts.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
+/*
```
Comment (georgewrmarshall) [inline jest.config.scripts.js:1]: Aligns with core

File: package.json [hunk 16]
```diff
@@ -1,93 +1,104 @@
 {
   "name": "@metamask/metamask-design-system",
   "version": "0.0.0",
+  "private": true,
```
Comment (georgewrmarshall) [inline package.json:4]: private is use to prevent the root repository being accidentally published

File: package.json [hunk 17]
```diff
@@ -1,93 +1,104 @@
...
+    "test:clean": "yarn workspaces foreach --all --parallel --verbose run test:clean && yarn test",
+    "test:packages": "yarn test:verbose --silent --collectCoverage=false --reporters=jest-silent-reporter",
+    "test:scripts": "yarn jest --config ./jest.config.scripts.js --silent",
+    "test:verbose": "yarn workspaces foreach --all --parallel --verbose run test:verbose",
+    "update-readme-content": "ts-node scripts/update-readme-content.ts",
+    "workspaces:list-versions": "./scripts/list-workspace-versions.sh"
+  },
+  "simple-git-hooks": {
+    "pre-push": "yarn lint"
+  },
```
Comment (georgewrmarshall) [inline package.json:42]: All these script should work with the exception of `prepare-preview-builds` and `publish-previews` which I have created an issue for because I still believe we will want this functionality in the near future https://github.com/MetaMask/metamask-design-system/issues/26

File: package.json [hunk 18]
```diff
@@ -1,93 +1,104 @@
...
+  "resolutions": {
+    "elliptic@6.5.4": "^6.5.7",
+    "fast-xml-parser@^4.3.4": "^4.4.1",
+    "ws@7.4.6": "^7.5.10"
   },
   "devDependencies": {
-    "@arethetypeswrong/cli": "^0.15.3",
+    "@babel/core": "^7.23.5",
```
Comment (georgewrmarshall) [inline package.json:49]: We can be certain that all these dependencies are needed because we use `depcheck` in our linting scripts, which ensures that all dependencies are being used.

Screenshot 2024-10-07 at 8 30 52 PM

File: packages/metamask-module-template/LICENSE [hunk 19]
```diff
@@ -0,0 +1,20 @@
+MIT License
```
Comment (georgewrmarshall) [inline packages/metamask-module-template/LICENSE:1]: Added using `create-package` script as an example package which we can remove in future

File: packages/metamask-module-template/README.md [hunk 20]
```diff
@@ -0,0 +1,15 @@
+# `@metamask/metamask-module-template`
```
Comment (georgewrmarshall) [inline packages/metamask-module-template/README.md:1]: All files within `metamask-module-template` were added using the `create-package` script as an example package to test that all package scripts work. We can remove this one as soon as we have another package.

File: release.config.json [hunk 21]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (georgewrmarshall) [inline release.config.json:1]: Is used with https://github.com/MetaMask/create-release-branch to determine our release strategy

File: scripts/create-package/README.md [hunk 22]
```diff
@@ -0,0 +1,4 @@
+# `create-package`
```
Comment (georgewrmarshall) [inline scripts/create-package/README.md:1]: Aligns with core to enable a CLI to create another package we will likely want to appropriate this core `create-component`

File: scripts/semver.sh [hunk 23]
```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env bash
```
Comment (georgewrmarshall) [inline scripts/semver.sh:1]: Used to compare versions by `migrate-tags` for migrating packages

File: scripts/since-latest-release.sh [hunk 24]
```diff
@@ -0,0 +1,145 @@
+#!/bin/bash
```
Comment (georgewrmarshall) [inline scripts/since-latest-release.sh:1]: This script is a useful tool for developers working on packages to quickly view changes since the last release, which can be helpful for changelog generation, code reviews, or general development tracking.

File: scripts/update-changelog.sh [hunk 25]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
```
Comment (georgewrmarshall) [inline scripts/update-changelog.sh:1]: Used with auto-changelog to update package changelogs

File: scripts/update-readme-content.ts [hunk 26]
```diff
@@ -0,0 +1,186 @@
+#!yarn ts-node
```
Comment (georgewrmarshall) [inline scripts/update-readme-content.ts:1]: - Retrieves workspace information
- Generates the dependency graph
- Generates the package list
- Updates the README with both

File: scripts/validate-changelog.sh [hunk 27]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
Comment (georgewrmarshall) [inline scripts/validate-changelog.sh:1]: Part of the CI/CD process to ensure that the changelog for a specific package is valid before allowing a release or merge

File: tsconfig.build.json [hunk 28]
```diff
@@ -1,24 +1,7 @@
...
-    "./src/**/__fixtures__/**/*",
-    "./src/**/__mocks__/**/*",
-    "./src/**/__test__/**/*",
-    "./src/**/__tests__/**/*",
-    "./src/**/__snapshots__/**/*",
-    "./src/**/*.test.ts",
-    "./src/**/*.test-d.ts",
-    "./src/**/*.test.*.ts"
-  ]
+  "references": [
```
Comment (georgewrmarshall) [inline tsconfig.build.json:2]: Aligns with core. Updates automatically with create-package script

File: tsconfig.json [hunk 29]
```diff
@@ -1,16 +1,6 @@
...
-    "module": "CommonJS",
-    "moduleResolution": "node",
-    "noEmit": true,
-    "noErrorTruncation": true,
-    "noUncheckedIndexedAccess": true,
-    "strict": true,
-    "target": "es2020"
-  },
-  "exclude": ["./dist", "**/node_modules"]
+  "compilerOptions": { "esModuleInterop": true, "noEmit": true },
```
Comment (georgewrmarshall) [inline tsconfig.json:2]: Aligns with core

File: tsconfig.packages.build.json [hunk 30]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (georgewrmarshall) [inline tsconfig.packages.build.json:1]: Aligns with core

File: tsconfig.packages.json [hunk 31]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (georgewrmarshall) [inline tsconfig.packages.json:1]: Aligns with core

File: tsconfig.scripts.json [hunk 32]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (georgewrmarshall) [inline tsconfig.scripts.json:1]: Aligns with core

File: yarn.config.cjs [hunk 33]
```diff
@@ -1,32 +1,328 @@
+// @ts-check
```
Comment (georgewrmarshall) [inline yarn.config.cjs:1]: Aligns with core but has been modified to be more generic when handling critical dependencies

File: .github/dependabot.yml [hunk 34]
```diff
@@ -1,15 +1,12 @@
 # Please see the documentation for all configuration options:
-# https://docs.github.com/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file
+# https://help.github.com/github/administering-a-repository/configuration-options-for-dependency-updates
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply .github/dependabot.yml]: Updated https://github.com/MetaMask/metamask-design-system/pull/17/commits/991fc6ea41712891f00f4181f50eb8dc4c224ad7

File: .yarnrc.yml (hunk 6 above)
Comment (georgewrmarshall) [reply .yarnrc.yml]: Updated https://github.com/MetaMask/metamask-design-system/pull/17/commits/991fc6ea41712891f00f4181f50eb8dc4c224ad7

File: .github/workflows/publish-preview.yml (hunk 1 above)
Comment (georgewrmarshall) [reply .github/workflows/publish-preview.yml:1]: Add a comment to the file that this script isn't working and to add a link to the issue "To do:"

File: docs/contributing.md [hunk 35]
```diff
@@ -0,0 +1,267 @@
...
+
+   4. Run `yarn install`.
+
+3. Due to the use of Yarn's `file:` protocol, if you update the package in the monorepo, then you'll need to run `yarn install` in the project again.
+
+### Testing changes to packages with preview builds
+
+If you want to test changes to a package where it would be unwieldy or impossible to use a local version, such as on CI, you can publish a preview build and configure your project to use it.
+
+#### Publishing preview builds as a MetaMask contributor
```
Comment (georgewrmarshall) [inline docs/contributing.md:90]: Add note that this script doesn't work and that we will be working on a fix and link the issue

File: docs/migrate-tags.md (hunk 11 above)
Comment (georgewrmarshall) [reply docs/migrate-tags.md:1]: Create a ticket to decide whether to remove migration scripts and documentation once we have migrated design-tokens

File: scripts/prepare-preview-builds.jq [hunk 36]
```diff
@@ -0,0 +1,9 @@
+# The name is overwritten, causing the package to get published under a
```
Comment (georgewrmarshall) [inline scripts/prepare-preview-builds.jq:1]: Add comment to fix

File: scripts/prepare-preview-builds.sh [hunk 37]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env bash
```
Comment (georgewrmarshall) [inline scripts/prepare-preview-builds.sh:1]: Add comment to fix

File: yarn.config.cjs [hunk 38]
```diff
@@ -1,32 +1,328 @@
...
+      // The root workspace (and only the root workspace) must specify the Yarn
+      // version required for development.
+      if (isChildWorkspace) {
+        workspace.unset('packageManager');
+      } else {
+        expectWorkspaceField(workspace, 'packageManager', 'yarn@4.2.2');
+      }
+
+      // All packages must specify a minimum Node.js version of 18.18.
+      expectWorkspaceField(workspace, 'engines.node', '^18.18 || >=20');
```
Comment (georgewrmarshall) [inline yarn.config.cjs:218]: @garrettbear - How do we stay up to date with yarn, node and other dependencies while being compatible with our consumers and making sure we align with Wallet Framework and other repos and MetaMask engineering teams?

**Review by brianacnguyen — APPROVED** [review]

**Review by garrettbear — APPROVED** [review]
🍀🍀🍀🍀
